### ISSUE TITLE
Release lock NPE on game launch

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#5963]
**How can the problem be recreated?**
- launched all AI game, single player, WWII v3 1941
- saw game crash after clicking 'start button'

**Which Engine Version are you using?** 

316733639f4c493a1bc8066ee54afdb01d42ccfc

**Additional Notes**
Appears to be a race condition, problem is not consistent, relaunched and was able to start a game without problem.

**Zipped Save Game**

None, crashed on game launch

** Stack Trace **
```
SEVERE: null
java.lang.NullPointerException
	at org.triplea.thread.LockUtil.releaseLock(LockUtil.java:80)
	at games.strategy.engine.data.GameData.releaseReadLock(GameData.java:343)
	at games.strategy.triplea.ui.EconomyPanel$ResourceTableModel.getRowCount(EconomyPanel.java:167)
	at java.desktop/javax.swing.JTable.getRowCount(JTable.java:2650)
	at java.desktop/javax.swing.plaf.basic.BasicTableUI.createTableSize(BasicTableUI.java:1732)
	at java.desktop/javax.swing.plaf.basic.BasicTableUI.getPreferredSize(BasicTableUI.java:1773)
	at java.desktop/javax.swing.JComponent.getPreferredSize(JComponent.java:1680)
	at java.desktop/javax.swing.ScrollPaneLayout.layoutContainer(ScrollPaneLayout.java:795)
	at java.desktop/java.awt.Container.layout(Container.java:1537)
	at java.desktop/java.awt.Container.doLayout(Container.java:1526)
	at java.desktop/java.awt.Container.validateTree(Container.java:1722)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validate(Container.java:1657)
	at java.desktop/java.awt.Container.validateUnconditionally(Container.java:1694)
	at java.desktop/java.awt.Window.show(Window.java:1050)
	at java.desktop/java.awt.Component.show(Component.java:1716)
	at java.desktop/java.awt.Component.setVisible(Component.java:1663)
	at java.desktop/java.awt.Window.setVisible(Window.java:1031)
	at games.strategy.engine.framework.startup.mc.HeadedLaunchAction.lambda$startGame$2(HeadedLaunchAction.java:69)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

**Comment by ron-murhammer** [comment]
@DanVanAtta Any more detail here? Not sure there is much that can be done here.

**Comment by DanVanAtta** [comment]
What kind of details would you be looking for here?

Race conditions are hard to repro, nonetheless the game *did* crash. This should be addressed in one way or another, we shouldn't let game crashing problems just pile up.

**Comment by ron-murhammer** [comment]
Just wondering if you've seen it multiple times or if there is something to make it more likely. If I can't reproduce it even inconsistently then there isn't really anything I can do.

**Comment by DanVanAtta** [comment]
1. Locking problems are difficult, it's one reason I think we should avoid using locks as it results in very difficult to trigger and diagnose race conditions.
2. There are tools available to us, analyze the code, try to artificially trigger the race condition with injected delays or constant condition hacks.
3. Analyzing the code we can probably determine what the problem is without necessarily understanding how it happens, which can lead to a potential fix.

A 'transient' problem, a race condition that does not happen often, is broken code all the same just as any other bug. The fact we made this problem rare and hard to fix is on us, it's not on the users to just experience game crashes because "it doesn't happen often" (or in other words, the code is not correct).

**Comment by DanVanAtta** [comment]
Observed this error again:
```
SEVERE: null
java.lang.NullPointerException
	at org.triplea.thread.LockUtil.releaseLock(LockUtil.java:80)
	at games.strategy.engine.data.GameData.releaseReadLock(GameData.java:343)
	at games.strategy.triplea.ui.StatPanel$StatTableModel.getRowCount(StatPanel.java:241)
```

Fresh instance of game, starting a local game on 1941 v3 with all AI players (I don't know if all AI players is related)

**Comment by DanVanAtta** [comment]
Got a thread dump, "Thread-6" seems suspicious:
```
"Thread-6" #38 prio=6 os_prio=0 cpu=330.40ms elapsed=130.75s tid=0x00007f7df40b1800 nid=0x75e waiting on condition  [0x00007f7dd5465000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.2/Native Method)
	- parking to wait for  <0x0000000710625d08> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(java.base@11.0.2/LockSupport.java:194)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(java.base@11.0.2/AbstractQueuedSynchronizer.java:885)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(java.base@11.0.2/AbstractQueuedSynchronizer.java:917)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@11.0.2/AbstractQueuedSynchronizer.java:1240)
	at java.util.concurrent.locks.ReentrantReadWriteLock$WriteLock.lock(java.base@11.0.2/ReentrantReadWriteLock.java:959)
	at org.triplea.thread.LockUtil.acquireLock(LockUtil.java:76)
	at games.strategy.engine.data.GameData.acquireWriteLock(GameData.java:354)
	at games.strategy.engine.history.HistoryWriter.closeCurrent(HistoryWriter.java:81)
	at games.strategy.engine.history.HistoryWriter.startNextStep(HistoryWriter.java:40)
	at games.strategy.engine.framework.ServerGame$1.stepChanged(ServerGame.java:145)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@11.0.2/Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@11.0.2/NativeMethodAccessorImpl.java:62)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@11.0.2/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@11.0.2/Method.java:566)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeSingle(EndPoint.java:146)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeMultiple(EndPoint.java:129)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeLocal(EndPoint.java:114)
```
AWT-EventQueue:
```
"AWT-EventQueue-0" #18 prio=6 os_prio=0 cpu=2311.30ms elapsed=140.65s tid=0x00007f7e6c852000 nid=0x708 waiting on condition  [0x00007f7dfc5bb000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.2/Native Method)
	- parking to wait for  <0x00000007056bc018> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(java.base@11.0.2/LockSupport.java:194)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@11.0.2/AbstractQueuedSynchronizer.java:2081)
	at java.awt.EventQueue.getNextEvent(java.desktop@11.0.2/EventQueue.java:566)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(java.desktop@11.0.2/EventDispatchThread.java:190)
	at java.awt.EventDispatchThread.pumpEventsForFilter(java.desktop@11.0.2/EventDispatchThread.java:124)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(java.desktop@11.0.2/EventDispatchThread.java:113)
	at java.awt.EventDispatchThread.pumpEvents(java.desktop@11.0.2/EventDispatchThread.java:109)
	at java.awt.EventDispatchThread.pumpEvents(java.desktop@11.0.2/EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.run(java.desktop@11.0.2/EventDispatchThread.java:90)
```

"2 similar threads (parking)"
```
"ForkJoinPool.commonPool-worker-3" #28 daemon prio=6 os_prio=0 cpu=14.07ms elapsed=131.14s tid=0x00007f7df454e000 nid=0x754 waiting on condition  [0x00007f7dd62aa000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.2/Native Method)
	- parking to wait for  <merged>(a java.util.concurrent.ForkJoinPool)
	at java.util.concurrent.locks.LockSupport.park(java.base@11.0.2/LockSupport.java:194)
	at java.util.concurrent.ForkJoinPool.runWorker(java.base@11.0.2/ForkJoinPool.java:1628)
	at java.util.concurrent.ForkJoinWorkerThread.run(java.base@11.0.2/ForkJoinWorkerThread.java:177)
```
 
ForkJoinPool.commonPool-worker-9 (parking)
```
"ForkJoinPool.commonPool-worker-9" #31 daemon prio=6 os_prio=0 cpu=14.33ms elapsed=131.14s tid=0x00007f7da4001000 nid=0x757 waiting on condition  [0x00007f7dd5fa7000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.2/Native Method)
	- parking to wait for  <0x0000000710648df8> (a java.util.concurrent.ForkJoinPool)
	at java.util.concurrent.locks.LockSupport.parkUntil(java.base@11.0.2/LockSupport.java:275)
	at java.util.concurrent.ForkJoinPool.runWorker(java.base@11.0.2/ForkJoinPool.java:1619)
	at java.util.concurrent.ForkJoinWorkerThread.run(java.base@11.0.2/ForkJoinWorkerThread.java:177)

```

**Comment by DanVanAtta** [comment]
Adding full stack trace, looks like the NPE is on AWT-eventqueue:
```
INFO: Adding endpoint: games.strategy.engine.message.unifiedmessenger.HasEndPointImplementor:games.strategy.engine.framework.AbstractGame.DISPLAY_CHANNEL, from: Node(name=dummy, address=dan-desk/127.0.1.1, port=0)
Mar 05, 2020 11:18:54 PM games.strategy.engine.message.UnifiedMessengerHub messageReceived
INFO: Adding endpoint: games.strategy.engine.message.unifiedmessenger.HasEndPointImplementor:games.strategy.engine.framework.AbstractGame.SOUND_CHANNEL, from: Node(name=dummy, address=dan-desk/127.0.1.1, port=0)
Mar 05, 2020 11:18:55 PM org.triplea.game.client.HeadedGameRunner lambda$initializeClientSettingAndLogging$0
SEVERE: null
java.lang.NullPointerException
	at org.triplea.thread.LockUtil.releaseLock(LockUtil.java:80)
	at games.strategy.engine.data.GameData.releaseReadLock(GameData.java:343)
	at games.strategy.triplea.ui.StatPanel$StatTableModel.getRowCount(StatPanel.java:241)
	at java.desktop/javax.swing.JTable.getRowCount(JTable.java:2650)
	at java.desktop/javax.swing.JTable.getCellRect(JTable.java:2921)
	at java.desktop/javax.swing.plaf.basic.BasicTableUI.createTableSize(BasicTableUI.java:1734)
	at java.desktop/javax.swing.plaf.basic.BasicTableUI.getPreferredSize(BasicTableUI.java:1773)
	at java.desktop/javax.swing.JComponent.getPreferredSize(JComponent.java:1680)
	at java.desktop/javax.swing.JViewport.getViewSize(JViewport.java:1119)
	at java.desktop/javax.swing.plaf.basic.BasicScrollPaneUI.syncScrollPaneWithViewport(BasicScrollPaneUI.java:330)
	at java.desktop/javax.swing.plaf.basic.BasicScrollPaneUI$Handler.stateChanged(BasicScrollPaneUI.java:1126)
	at java.desktop/javax.swing.JViewport.fireStateChanged(JViewport.java:1490)
	at java.desktop/javax.swing.JViewport.reshape(JViewport.java:954)
	at java.desktop/java.awt.Component.setBounds(Component.java:2326)
	at java.desktop/java.awt.Component.setBounds(Component.java:2477)
	at java.desktop/javax.swing.ScrollPaneLayout.layoutContainer(ScrollPaneLayout.java:888)
	at java.desktop/java.awt.Container.layout(Container.java:1537)
	at java.desktop/java.awt.Container.doLayout(Container.java:1526)
	at java.desktop/java.awt.Container.validateTree(Container.java:1722)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validateTree(Container.java:1731)
	at java.desktop/java.awt.Container.validate(Container.java:1657)
	at java.desktop/java.awt.Container.validateUnconditionally(Container.java:1694)
	at java.desktop/java.awt.Window.show(Window.java:1050)
	at java.desktop/java.awt.Component.show(Component.java:1716)
	at java.desktop/java.awt.Component.setVisible(Component.java:1663)
	at java.desktop/java.awt.Window.setVisible(Window.java:1031)
	at games.strategy.engine.framework.startup.mc.HeadedLaunchAction.lambda$startGame$2(HeadedLaunchAction.java:70)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

**Comment by DanVanAtta** [comment]
Looked into this, `LockUtil` has some very suspicious code as well with non-atomic updates to lock counts. EG:

```
    if(!locksHeld.get().containsKey(lock)) {
      final int current = locksHeld.get().get(lock);
      locksHeld.get().put(lock, current + 1);
    }
```

If we fail to acquire a lock, the NPE is thrown when trying to release the lock.  I was not able to simulate this situation, though it could very well be possible perhaps. We could do a blind fix where releasing a lock that was never obtained in the first place is a no-op, we could fail sooner and fail if we do not acquire a lock (in which case we would still have this problem, but hopefully easier to track down), or we could try to add some sort of blocking waiting to acquire a lock (which could introduce concurrency problems). Yet another option is to try and remove dependencies on locks, for examle we grab locks whenever moving the mouse and scrolling over territories, set copies and using thread safe data structures perhaps are better option now that were not available when this game was first written under Java 1.4

**Comment by RoiEXLab** [comment]
@DanVanAtta The NPE that is observed here appears to occur when trying to release a lock that isn't held: The Map returns null for the given lock -> NPE when trying to cast it to `int`.
We could easily fix that by adding a check, but as you already mentioned maps are not really known for their atomic behaviour. Using additional mutexes or AtomicIntegers instead of ints, might help a bit, but I'm not really convinced the resulting code will be a lot better.
I stumbled onto this part of the code before and I would like to see it gone to be honest.

I did some research and it appears that the underlying problem this class tries to partially solve still exists.
I believe that if we just dropped the code, everything would be fine for the near future, but it's possible that we'll eventually run into a deadlock at some point, simply because the code is so complex so we might want to have some "real protection" against it.
I did some further research and there are a couple of things we could do with few lines of code.
1. Never lock indefinitely. Not the cleanest solution on it's own, but if we only ever lock with let's say a minute long timeout we will never run into any deadlock (using this code obviously) and we can just throw an exception that an operation took to long in this case, if we get a lot of those we can either increase the timeout, or investigate the locks.
2. If option number 1 is not safe enough to detect a deadlock, we could try something fancy: Once we get a lock timeout we could invoke [ThreadMXBean#findDeadlockedThreads()](https://docs.oracle.com/en/java/javase/11/docs/api/java.management/java/lang/management/ThreadMXBean.html#findDeadlockedThreads()) to check if a deadlock appears to be the problem and inform the user that they should create a bug report.

For option 2 I have somthing like this in mind:
```java
static void tryAcquiringLock(final Lock lock) {
  if (!lock.tryLock(1, TimeUnit.MINUTES)) {
    Thread lockedThread = Thread.currentThread();
    new Thread(() -> {
      final int MAX_TRIES = 1000;
      int count = 0;
      // Wait until thread is re-locked again.
      while (lockedThread.getState() != Thread.State.BLOCKED && count < MAX_TRIES) {
        // Throws InterruptedException, but you know the drill
        // we just ignore that for now
        Thread.sleep(100);
        count++;
      }
      if (lockedThread.getState() == Thread.State.BLOCKED) {
        final int[] deadlockedThreads = ThreadMXBean.findDeadlockedThreads();
        if (deadlockedThreads != null) {
          for (final ThreadInfo info : ThreadMXBean.getThreadInfo(deadlockedThreads, Integer.MAX_VALUE)) {
            final Exception e = new Exception("Deadlock!");
            e.setStackTrace(info.getStackTrace())
            logger.error("Thread " + info.getThreadName() + " deadlocked.", e);
          }
        }
      }
    }).start();
    lock.lock();
  }
}
```
